internal/kgateway/extensions2/plugins/listenerpolicy/listener_plugin.go

```
@@ -0,0 +1,247 @@
+				// create empty policy report entry with no ancestor refs
+				rp.Policy(policyKey, 0)
+			}
+		}
+	}
+	return sdk.Plugin{
+		ExtraHasSynced: col.HasSynced,
+		ContributesPolicies: map[schema.GroupKind]sdk.PolicyPlugin{
+			kgwwellknown.ListenerPolicyGVK.GroupKind(): {
+				NewGatewayTranslationPass:       NewGatewayTranslationPass,
+				Policies:                        policyCol,
+				ProcessPolicyStaleStatusMarkers: processMarkers,
+				GetPolicyStatus:                 getPolicyStatusFn(cli),
+				PatchPolicyStatus:               patchPolicyStatusFn(cli),
+			},
+		},
+	}
+}
+
+func NewGatewayTranslationPass(tctx ir.GwTranslationCtx, reporter reporter.Reporter) ir.ProxyTranslationPass {
+	return &listenerPolicyPluginGwPass{
+		reporter: reporter,
+	}
+}
+
+func (p *listenerPolicyPluginGwPass) Name() string {
+	return "listenerpolicy"
+}
+
+func (p *listenerPolicyPluginGwPass) ApplyListenerPlugin(
+	pCtx *ir.ListenerContext,
+	out *envoylistenerv3.Listener,
+) {
+	logger.Debug("applying to listener", "listener", out.Name, "policyType", fmt.Sprintf("%T", pCtx.Policy))
+	pol, ok := pCtx.Policy.(*listenerPolicy)
+	if !ok || pol == nil {
+		logger.Warn("policy is not listenerPolicy type or is nil", "ok", ok, "pol", pol)
+		return
+	}
+
+	logger.Debug("listenerPolicy found", "proxy_protocol", pol.proxyProtocol)
+	// Add proxy protocol listener filter if configured
+	if pol.proxyProtocol != nil {
+		p.applyProxyProtocol(out, pol.proxyProtocol)
+	}
+}
+
+func convertProxyProtocolConfig(objSrc ir.ObjectSource, config *v1alpha1.ProxyProtocolConfig) *anypb.Any {
+	if config == nil {
+		return nil
+	}
+	// Create the proxy protocol configuration
+	proxyProtocolConfig := &proxy_protocol.ProxyProtocol{
+		StatPrefix: fmt.Sprintf("%s_%s", objSrc.Namespace, objSrc.Name),
+	}
+
+	// Marshal to Any
+	proxyProtocolAny, err := utils.MessageToAny(proxyProtocolConfig)
+	if err != nil {
+		logger.Error("failed to marshal proxy protocol config",
+			"error", err)
+	}
```

We could also return this error and propagate to the IR's Error field. AFAIK, we don't handle errors on the IR for the HLP IR though.

---

it should never happen, i thought to panic here but ended up doing the log. maybe i should add a panic